radiacode/radiacode.py

```diff
@@ -138,10 +139,25 @@ def write_request(self, command_id: int | VSFR, data: Optional[bytes] = None) ->
         assert retcode == 1
         assert r.size() == 0
 
-    def batch_read_vsfrs(self, vsfr_ids: list[VSFR]) -> list[int]:
+    def batch_read_vsfrs(self, vsfr_ids: list[VSFR], unpack_format=None) -> list[int]:
+        """Read multiple VSFRs
+
+        Args:
+            vsfr_ids: a list of VSFRs to fetch
+            unpack_format: a `struct` format string used to unpack the response
+                into certain data types, eg. '<4x3f` to skip the first 4 bytes,
+                and interpret the remaining bytes into 3 floats.
+
+                If not given, the response is simply decoded into a list of ints.
+        """
         assert len(vsfr_ids)
-        r = self.execute(COMMAND.RD_VIRT_SFR_BATCH, b''.join(struct.pack('<I', int(c)) for c in vsfr_ids))
-        ret = [r.unpack('<I')[0] for _ in range(len(vsfr_ids))]
+        msg = [struct.pack('<I', len(vsfr_ids))]
+        msg.extend([struct.pack('<I', int(c)) for c in vsfr_ids])
+        r = self.execute(COMMAND.RD_VIRT_SFR_BATCH, b''.join(msg))
+        if unpack_format:
+            ret = r.unpack(unpack_format)
+        else:
+            ret = [r.unpack('<I')[0] for _ in range(len(vsfr_ids) + 1)]
```

Comment:
Why `+1` and 'skip first 4 bytes' in `'<4x3f` to skip the first 4 bytes,`?
What's inside those 4 bytes?

Reply:
Even if it's unknown at the moment, I think it's better to explicitly read it and ignore it (with a comment in the code), and not take it into account in `unpack_format ` / return array values

Reply:
the first 4 bytes are a bitfield that indicates whether the requested VSFR is valid. in the example, I requested 3 VSRFs (the ones containing the energy calibration), and those are valid, so 3 bits are set. 

the point of `unpack_format` is to allow users to build their own queries in whatever way they want. `<4x3f` was simply an example; `<I3f` or `<Ifff` are just as valid.

Reply:
In that case, maybe it's better to unpack only 4 bytes in `batch_read_vsfrs`, parse it (to list or tuple of bools) and return it with raw response - if we can't distinguish replies to different requests of batch in this function, then I think it's better to construct and parse data on the caller's side.
So, my proposal is:
```python
def batch_read_vsfrs(self, vsfr_ids: list[VSFR]) -> tuple[list[bool], BytesBuffer]
```